### PR TITLE
Move colonizer CanColonize location condition to FOCS. Allow outposts…

### PR DESCRIPTION
**Pull request by agrrr3**
… to be build even if not CanColonize
[forum discussion](https://freeorion.org/forum/viewtopic.php?t=12849)

### CODE REVIEW

**Comment by Blue-Award** [comment]
My understanding is AI should be fine, let me repeat what I wrote on the forum:

Looks like AI will assume ship is an outposter regardless of its species, so should be good on that front:

https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/default/python/AI/FleetUtilsAI.py#L438-L445

(backend has some vestiges of Hulls being potentially able to have colony capacity and sums that with parts capacity; AI just does parts capacity... but hull capacity is just  hardcoded as 0.0 so I guess not to worry, eh?)

I went into the rabbit hole of production side of this. I understand this boils down to the following, so should also be good:

https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/default/python/AI/ship_design/_ship_design_cache.py#L308-L333 (specifically the line https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/default/python/AI/ship_design/_ship_design_cache.py#L331 - cf. comment on https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/default/python/AI/ShipDesignAI.py#L609 )

It's how the ProductionAI takes info from that cache: https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/default/python/AI/ProductionAI.py#L1090

**Comment by Blue-Award** [comment]
Aha I think you covered the building part but not the colonization part. I think it is necessary to change Ship.cpp CanColonize so it does not check its species 

https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/universe/Ship.cpp#L272-L280

**Comment by agrrr3** [comment]
> My understanding is AI should be fine, let me repeat what I wrote on the forum:
> (specifically the line
> 
> https://github.com/freeorion/freeorion/blob/3ea54a4bc626f2425f7f34da6f19802833602c39/default/python/AI/ship_design/_ship_design_cache.py#L331

yeah that looks like it builds a list of available stuff per planet and probably chooses best suited from that in a later step(?)

> Aha I think you covered the building part but not the colonization part. I think it is necessary to change Ship.cpp CanColonize so it does not check its species

sounds right, will fix later

File: universe/Ship.cpp [hunk 1]
```diff
@@ -272,11 +272,18 @@ bool Ship::HasFighters(const Universe& universe) const {
...
     if (m_species_name.empty())
         return false;
-    const Species* species = sm.GetSpecies(m_species_name);
-    if (!species || !species->CanColonize())
-        return false;
     const ShipDesign* design = universe.GetShipDesign(m_design_id);
-    return design && design->CanColonize(); // use design->CanColonize because zero-capacity colony ships still count as outpost ships, can "can colonize" as far as order / the UI are concerned
+    if (!design)
+        return false;
+    if (design && design->CanColonize()) {
```
Comment (Blue-Award) [inline universe/Ship.cpp:276]: due to earlier "if (!design) return false" check, design here should always be true so no need to check second time. Or conversely drop the earlier check and count on the final "return false"

**Comment by Grummel7** [comment]
The change seems to work. I even saw an AI build an outpost ship on a planet that couldn't build colony ships. The AI currently seems to have a serious problem with outpost ships: it doesn't send them anywhere. But this not caused by this PR, I observe the same with master.

Regarding this PR I think it would be good to rename some function. With this change we have two meanings for CanColonize. For a species it means can build colonies, for a ship/design it means can build a colony or an outpost. Or perhaps just comment them better (has colonisation parts is not so helpful, say can build a colony or outpost and add it to the same function in ship as well)

**Comment by agrrr3** [comment]
> Regarding this PR I think it would be good to rename some function. With this change we have two meanings for CanColonize.

outposting and colonizing, like before? CanColonizeOrOutpost seems klunky. 

> For a species it means can build colonies, for a ship/design it means can build a colony or an outpost. Or perhaps just comment them better (has colonisation parts is not so helpful, say can build a colony or outpost and add it to the same function in ship as well)

just suggest somethinng

**Comment by geoffthemedio** [comment]
@Grummel7 @agrrr3 poke

**Comment by Blue-Award** [comment]
heeyyyy those gis guf and khakturians demand outposting rights!

**Comment by agrrr3** [comment]
If grummel does not come up with something better, i suggest we keep the
current CanColonize

**Pull request merged**